### PR TITLE
Feat: moved comments from pooling to ws

### LINKED ISSUE #111: move comments to websockets
current implementation just keeps listing comments every 5 seconds which is pretty annoying

https://github.com/briefercloud/briefer/blob/a341ecd326aad95dd2caf75e6372d79719b8347e/apps/web/src/hooks/useComments.ts#L26

### PR DESCRIPTION
**Pull request by hiteshwadhwani**
closes #111

### CODE REVIEW

**Review by defihook — COMMENTED** [review]
looks awesome

**Review by lucasfcosta — COMMENTED** [review]
Amazing work @hiteshwadhwani! I just had a first pass at this and I think it looks great.

I left a few minor comments to clean it up and also fixed a compilation error.

I also tested this on my machine and it works great.

I won't approve this just yet because I want @vieiralucas to double-check the authorization side of things before merging. We just need to make sure that people can only subscribe to docs/join rooms that belong to workspaces to which they have access.

Once that's done we can just address the minor points and merge.

File: apps/api/src/websocket/workspace/comments.ts [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    try {
+      const parsedPyload = Payload.safeParse(data)
+      if (!parsedPyload.success && callback) {
+        callback(getCallbackErrorResponse('Invalid payload'))
+        return
+      }
+      const { documentId, content } = parsedPyload.data as z.infer<
+        typeof Payload
+      >
+      // save to db
```
[the remark this answers is not in the document]
Comment (hiteshwadhwani) [reply apps/api/src/websocket/workspace/comments.ts]: done

File: apps/api/src/websocket/workspace/comments.ts [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+          userId: session.user.id,
+        },
+      })
+
+      const response: Comment = {
+        ...comment,
+        user: { name: session.user.name, picture: session.user.picture },
+      }
+
+      // ack to the user
```
[the remark this answers is not in the document]
Comment (hiteshwadhwani) [reply apps/api/src/websocket/workspace/comments.ts]: removed

File: apps/api/src/websocket/workspace/comments.ts [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+      callback?.(getCallbackSuccessResponse())
+
+      // broadcast to all users in the document room
```
[the remark this answers is not in the document]
Comment (hiteshwadhwani) [reply apps/api/src/websocket/workspace/comments.ts]: removed

File: apps/api/src/websocket/workspace/comments.ts [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+      const parsedData = z.object({ docId: z.string() }).safeParse(data)
+      if (!parsedData.success) {
+        callback && callback(getCallbackErrorResponse('Invalid payload'))
+        return
+      }
+      const roomId = getDocumentRoomId(parsedData.data.docId)
+      if (!socket.rooms.has(roomId)) {
+        await socket.join(roomId)
+      }
+      await emitWorkspaceDocumenttComments(socket, parsedData.data.docId)
```
[the remark this answers is not in the document]
Comment (hiteshwadhwani) [reply apps/api/src/websocket/workspace/comments.ts]: removed

File: apps/api/src/websocket/workspace/comments.ts [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+const broadcastComment = (
+  io: IOServer,
+  documentId: string,
+  comment: Comment
+) => {
+  io.to(getDocumentRoomId(documentId)).emit('workspace-comment', [comment])
+}
+
+const emitWorkspaceDocumenttComments = async (
```
[the remark this answers is not in the document]
Comment (hiteshwadhwani) [reply apps/api/src/websocket/workspace/comments.ts]: done

File: apps/web/src/hooks/useComments.ts [hunk 6]
```diff
@@ -1,51 +1,45 @@
...
-  )
+  const [comments, setComments] = useState<Comment[]>([])
+  const [error, setError] = useState<Error>(null)
+  const socket = useWebsocket()
 
-  const comments = useMemo(() => data ?? [], [data])
+  useEffect(() => {
+    socket?.emit("join-workspace-document", {docId}, (ack: CommentAck) => {
+      if(ack.status === 'error'){
+        console.log({ack})
```
[the remark this answers is not in the document]
Comment (hiteshwadhwani) [reply apps/web/src/hooks/useComments.ts]: removed

File: apps/api/src/websocket/index.ts [hunk 7]
```diff
@@ -160,6 +167,9 @@ export function createSocketServer(server: http.Server): Server {
       trackWork(refreshDataSource(io, socket, session))
     )
 
+    // expecting a workspace id and comment
```
[the remark this answers is not in the document]
Comment (hiteshwadhwani) [reply apps/api/src/websocket/index.ts]: removed

File: apps/web/src/components/Comments.tsx [hunk 8]
```diff
@@ -85,49 +85,55 @@ export default function Comments(props: Props) {
...
+                            <div className="flex items-center justify-center relative h-5 w-5 flex-none bg-gray-50 rounded-full">
+                              <UserIcon className="h-4 w-4" />
+                            </div>
+                          )}
+                          <span className="text-xs font-medium text-gray-900">
+                            {comment.user.name}
+                          </span>{' '}
+                        </div>
+                        <time
+                          dateTime={comment.createdAt}
```
[the remark this answers is not in the document]
Comment (hiteshwadhwani) [reply apps/web/src/components/Comments.tsx]: done

**Comment by hiteshwadhwani** [comment]
@lucasfcosta I have addressed the PR comments. Let me know if we need any changes from the auth side 👍

**Comment by vieiralucas** [comment]
Thank you @hiteshwadhwani, sorry that it took me some time to review.

**Pull request merged**